File: Ruby_version/Napakalaki/lib/specific_bad_consequence.rb
# encoding: utf-8
# begin specific_bad_consequence.rb

module NapakalakiGame
  class SpecificBadConsequence < BadConsequence
    
    def initialize(t, l, v, h)
      super(t, l)
      @specificVisibleTreasures = v
      @specificHiddenTreasures = h
    end
    
    def getSpecificVisibleTreasures
      @specificVisibleTreasures 
    end

    def getSpecificHiddenTreasures
      @specificHiddenTreasures 
    end
    
    def isEmpty
      @specificVisibleTreasures.empty? and @specificHiddenTreasures.empty?
    end
    
    def substractVisibleTreasure(t)
      if !@specificVisibleTreasures.empty? and @specificVisibleTreasures.include?(t)
        @specificVisibleTreasures.delete_at(@specificVisibleTreasures.index(t))
      end
    end
    
    def substractHiddenTreasure(t)
      if !@specificHiddenTreasures.empty? and @specificHiddenTreasures.include?(t)
        @specificHiddenTreasures.delete_at(@specificHiddenTreasures.index(t))
      end
    end
    
    def adjustToFitTreasureLists(v, h)
      vTType = v.collect {|t| t.getType}
      hTType = h.collect {|t| t.getType}
      vTFinal = (vTType & @specificVisibleTreasures).flat_map { |n| [n]*[vTType.count(n), @specificVisibleTreasures.count(n)].min }
      hTFinal = (hTType & @specificHiddenTreasures).flat_map { |n| [n]*[hTType.count(n), @specificHiddenTreasures.count(n)].min }
      bC = SpecificBadConsequence.new(@text, @levels, vTFinal, hTFinal)
    end
    
    def to_s
      super + "\n Tesoros visibles perdidos: #{@specificVisibleTreasures}\n Tesoros ocultos perdidos: #{@specificHiddenTreasures}"
    end
    
    def malRolloString
      super + "\n Tesoros visibles perdidos: #{@specificVisibleTreasures}\n Tesoros ocultos perdidos: #{@specificHiddenTreasures}"
    end
    
  end
end

=begin
  def adjustToFitTreasureLists(v, h) 
    # Primero borramos los tesoros que no están en ninguno
    vTType.delete_if {|t| !t.in?(@specificVisibleTreasures)}
    hTType.delete_if {|t| !t.in?(@specificHiddenTreasures)}

    # Miramos en los tipos unicos
    vTType.uniq.each do |type|
     # Adjustamos el número
      n1 = vTType.count(type)
      n2 = @specificVisibleTreasures.count(type)
      while n1 > n2 do
        vTType.delete(type)
        n1 -= 1
      end
    end

    # Miramos en los tipos unicos
    hTType.uniq.each do |type|
     # Adjustamos el número
     n1 = hTType.count(type)
     n2 = @specificHiddenTreasures.count(type)
     while n1 > n2 do
       hTType.delete(type)
       n1 -= 1
     end
  end
=end

# end specific_bad_consequence.rb

File: Java_version/Napakalaki/src/napakalaki/SpecificBadConsequence.java
// begin SpecificBadConsequence.java
package napakalaki;

// Usar arrays
import java.util.ArrayList;

public class SpecificBadConsequence extends BadConsequence {
    // Tesoros visibles específicos perdidos
    private ArrayList<TreasureKind> specificVisibleTreasures = new ArrayList();
    // Tesoros ocultos específicos perdidos
    private ArrayList<TreasureKind> specificHiddenTreasures = new ArrayList();
    
    // Constructor tipo 2: tesoros específicos
    public SpecificBadConsequence(String t, int l, ArrayList<TreasureKind> v,
                               ArrayList<TreasureKind> h) {
        super(t, l);
        this.specificVisibleTreasures = v;
        this.specificHiddenTreasures = h;
    }
    
    public ArrayList<TreasureKind> getSpecificHiddenTreasures() {
        return specificHiddenTreasures;
    }
    
    public ArrayList<TreasureKind> getSpecificVisibleTreasures() {
        return specificVisibleTreasures;
    }
    
    @Override
    public boolean isEmpty() {
        return specificHiddenTreasures.isEmpty() &&
                specificVisibleTreasures.isEmpty();
    }
    
    @Override
    public void substractVisibleTreasure(Treasure t) {
        if (!specificVisibleTreasures.isEmpty())
            specificVisibleTreasures.remove(t.getType());
    }
    
    @Override
    public void substractHiddenTreasure(Treasure t) {
        if (!specificHiddenTreasures.isEmpty())
            specificHiddenTreasures.remove(t.getType());
    }
    
    @Override
    public BadConsequence adjustToFitTreasureLists(ArrayList<Treasure> v,
            ArrayList<Treasure> h) {
        
        ArrayList<TreasureKind> vTType = new ArrayList();
        ArrayList<TreasureKind> hTType = new ArrayList();
        for (Treasure t: v)
            vTType.add(t.getType());
        for (Treasure t: h)
            hTType.add(t.getType());

        vTType.retainAll(specificVisibleTreasures);
        hTType.retainAll(specificHiddenTreasures);
        
        BadConsequence bd = new SpecificBadConsequence(getText(), getLevels(), 
                    vTType, hTType);
        return bd;
    }
    
    @Override
    public String toString() {
        return super.toString() + "\n Tesoros visibles perdidos: " +
                specificVisibleTreasures.toString() +
                "\nTesoros ocultos perdidos: " +
                specificHiddenTreasures.toString();
    }
    
    @Override
    public String toDiscard() {
        return "Visibles: " + specificVisibleTreasures.toString() +
            "\n Ocultos: " + specificHiddenTreasures.toString();
    }
}

// end SpecificBadConsequence.java